File: SearchInRotatedSortedArray.py
"""
Search in Rotated sorted array

Brute force approach -- Linear search - TC - O(n) SC - O(1)

Binary Search Approach --
TC - O(log n)
SC - O(1)
"""

nums = [4, 5, 6, 7, 0, 1, 2]
target = 0


class Solution:
    def search(self, nums, target):
        if nums is None or len(nums) == 0 or target is None: return -1

        n = len(nums)

        low = 0
        high = n - 1

        while low <= high:
            mid = low + (high - low) // 2

            if nums[mid] == target: return mid

            # check which part is sorted
            # left is sorted
            if nums[low] <= nums[mid]:
                # check if target is present
                if nums[low] <= target < nums[mid]:
                    high = mid - 1
                else:
                    low = mid + 1
            else:
                # right is sorted
                if nums[mid] <= target <= nums[high]:
                    low = mid + 1
                else:
                    high = mid - 1
        return -1


obj = Solution()
print(obj.search(nums, target))

File: old-sub/SearchInRotatedSortedArray.java
/*
Time Complexity: 0(logn)
Space Complexity: 0(1)
Approach: Perform binary search on the rotated sorted array and search conditionally.
 */
class Solution {

    public int search(int[] nums, int target) {

        int low = 0;
        int high = nums.length-1;
        int mid;

        while (low <= high){

            mid = (high-low)/2 + low;

            // System.out.println("low = "+low+ " mid = "+mid+ " high = "+high);

            if (nums[mid] == target){
                return mid;
            }

            else if (nums[mid] > target){

                if (nums[low] <= nums[mid] && target < nums[low]) {
                    // sorted array, chk target on rhs
                    low = mid + 1;
                    continue;
                }
                // target on lhs
                high = mid - 1;
            }

            else{

                if (nums[mid] <= nums[high] && target > nums[high]) {
                    // sorted array, chk target on lhs
                    // case [5,1,3] tgt=5
                    high = mid - 1;
                    continue;
                }
                // target on rhs
                low = mid + 1;
            }

        }
        return -1;
    }
}